src/logic/equiv/set.lean

```diff
@@ -512,6 +512,34 @@ begin
   by_cases hi : p i; simp [hi]
 end
 
+/-- An equation in the target type gives an equivalence between the corresponding preimages. -/
+@[simps]
+def preimage_congr {α β : Type*} (f : α → β)
+  {s s' : set β} (h : s = s') : f ⁻¹' s ≃ f ⁻¹' s' :=
+equiv.set_congr $ h ▸ rfl
+
+lemma preimage_congr_apply
+  {α β γ : Type*} {f : α → γ} {g : β → γ} (e : Π c, (f ⁻¹' {c}) ≃ (g ⁻¹' {c}))
+  {c c' : γ} (h : c = c') (x : f ⁻¹' {c'}) :
+  (preimage_congr g (congr_arg singleton h)) (e c ⟨x.1, begin simpa [h] using x.2, end⟩) = e c' x :=
```

Comment:
Probably better to use a coercion of `x` instead of `x.1`. Also may as well golf out the `simpa`
```suggestion
  preimage_congr g (congr_arg singleton h) (e c ⟨x, h.symm ▸ x.prop⟩) = e c' x :=
```

Reply:
Weirdly replacing the `.1`s with coercions in the `def` below breaks the proof. It doesn't cause trouble there though.